onnx/defs/math/defs.cc

```diff
@@ -1665,4 +1665,126 @@ ONNX_OPERATOR_SET_SCHEMA(
+  if (numInputs < 2 || !hasNInputShapes(ctx, static_cast<int>(numInputs))) {
+    return;
+  }
+
+  auto* output_shape = getOutputShape(ctx, 0);
+  std::string  left_equation;
+
+  auto mid_index = equation.find("->");
+  if (mid_index != std::string::npos) {
+    // Separate right and left hand sides of the equation
+    left_equation = equation.substr(0, mid_index);
+  } else {
+    // No right hand side
+    left_equation = equation;
+  }
+
+  std::string term;
+  size_t num_operands = 0;
+  size_t num_ellipsis = 0;
+  size_t num_ellipsis_indices = 0;
+  std::stringstream str(left_equation);
+
+  while(std::getline(str, term, ',')) {
+    auto ellipsis_index = term.find("...");
+    if (ellipsis_index != std::string::npos) {
+      // If there is an ellipsis, the number of dimensions it represents must be total dim - letter dimensions
+      size_t rank = ctx.getInputType(num_operands)->tensor_type().shape().dim_size();
+      if (num_ellipsis == 0) {
+        num_ellipsis_indices = rank - term.size() + 3;
+      } else { // ellipsis has been seen before. Check that if dimensions are compatible
+        if (num_ellipsis_indices != rank - term.size() + 3) {
+          fail_shape_inference("Ellipsis represents incompatible dimensions.");
+        }
+      }
+      num_ellipsis++;
+    }
+    num_operands++;
+  }
+
+  if (numInputs - 1 != num_operands) {
+    fail_shape_inference("Number of input tensors does not match the operand in the equation.");
+  }
+
+  const size_t number_of_letters = 26;
+  size_t num_letter_occurrences[number_of_letters] = {0};
+  if (mid_index != std::string::npos) {  // parse the user provided right hand side
+    std::string right_equation = equation.substr(mid_index+2);
+    auto right_ellipsis_index = right_equation.find("...");
+    if (right_ellipsis_index != std::string::npos) { // right hand side contains ellipsis
+      for (size_t i = 0; i < num_ellipsis; ++i) {
+        output_shape->add_dim();
+      }
+    }
+    for (char c: right_equation) { // add a dimension per each character in right hand equation
+      if (c != '.') {
+        output_shape->add_dim();
+      }
+    }
+  } else { // infer the dimension for right hand side of the equation
+    // If there's an ellipsis, add it's corresponding dimensions
+    for (size_t i = 0; i < num_ellipsis_indices; i++) {
+      output_shape->add_dim();
+    }
+    for (size_t i = 0; i < left_equation.size(); i++) { // count chars that appear exactly once on left hand side
+      if ((left_equation.at(i) != ',') && (left_equation.at(i) != '.')) {
+        num_letter_occurrences[left_equation.at(i) - 'a']++;
+      }
+    }
+    for (size_t index = 0; index < number_of_letters; index++) {
+      if (num_letter_occurrences[index] == 1) {
+        output_shape->add_dim();
+      }
+    }
+  }
+}
+
+static const char* Einsum_ver11_doc = R"DOC(
+The Einsum operator evaluates algebraic tensor operations on the operands, using the Einstein summation convention.).
```

Comment:
")." -> typo?

Reply:
I think it would be helpful to add a precise, complete, and unambiguous description of this language. From the description I have seen in other frameworks, there are some choices and ambiguities in what Einstein summation notation means and allows.

Reply:
It would be good to have a simple grammar for the language too. For example, are white-spaces allowed in the string? It is a good idea to allow spaces in the string and skip them. When I look at the rank-inference logic, the use term.size() makes me wonder if this assumes that there are no white-spaces in the string.

Reply:
Thanks @gramalingam, I've updated the doc with more details about equation string.